main.cpp

```diff
@@ -104,6 +108,406 @@ good to go!
+    FloatType& divide(float rhs);
+
+    FloatType& add(const FloatType& rhs);
+    FloatType& subtract(const FloatType& rhs);
+    FloatType& multiply(const FloatType& rhs);  
+    FloatType& divide(const FloatType& rhs);
+
+    FloatType& add(const DoubleType& rhs);
+    FloatType& subtract(const DoubleType& rhs);
+    FloatType& multiply(const DoubleType& rhs);  
+    FloatType& divide(const DoubleType& rhs);
+
+    FloatType& add(const IntType& rhs);
+    FloatType& subtract(const IntType& rhs);
+    FloatType& multiply(const IntType& rhs);  
+    FloatType& divide(const IntType& rhs);
+};
+
+// ---
+
+struct DoubleType
+{
+    double* value = nullptr;
+
+    DoubleType(double initValue) : value(new double(initValue)) {}
+    ~DoubleType()
+    {
+        delete value;
+        value = nullptr;
+    }
+
+    DoubleType& add(double rhs);
+    DoubleType& subtract(double rhs);
+    DoubleType& multiply(double rhs);
+    DoubleType& divide(double rhs);
+
+    DoubleType& add(const FloatType& rhs);
+    DoubleType& subtract(const FloatType& rhs);
+    DoubleType& multiply(const FloatType& rhs);
+    DoubleType& divide(const FloatType& rhs);
+
+    DoubleType& add(const DoubleType& rhs);
+    DoubleType& subtract(const DoubleType& rhs);
+    DoubleType& multiply(const DoubleType& rhs);
+    DoubleType& divide(const DoubleType& rhs);
+
+    DoubleType& add(const IntType& rhs);
+    DoubleType& subtract(const IntType& rhs);
+    DoubleType& multiply(const IntType& rhs);
+    DoubleType& divide(const IntType& rhs);
+};
+
+// ---
+
+struct IntType
+{
+    int* value = nullptr;
+
+    IntType(int initValue) : value(new int(initValue)) {}
+    ~IntType()
+    {
+        delete value;
+        value = nullptr;
+    }
+
+
+    IntType& add(int rhs);
+    IntType& subtract(int rhs);
+    IntType& multiply(int rhs);
+    IntType& divide(int rhs);
+
+    IntType& add(const FloatType& rhs);
+    IntType& subtract(const FloatType& rhs);
+    IntType& multiply(const FloatType& rhs);
+    IntType& divide(const FloatType& rhs);
+
+    IntType& add(const DoubleType& rhs);
+    IntType& subtract(const DoubleType& rhs);
+    IntType& multiply(const DoubleType& rhs);
+    IntType& divide(const DoubleType& rhs);
+
+    IntType& add(const IntType& rhs);
+    IntType& subtract(const IntType& rhs);
+    IntType& multiply(const IntType& rhs);
+    IntType& divide(const IntType& rhs);
+};
+
+// ---------
+
+
+FloatType& FloatType::add(float rhs) 
+{
+    if (value != nullptr) *value += rhs;
+    return *this;
+}
+
+FloatType& FloatType::subtract(float rhs) 
+{
+    if (value != nullptr) *value -= rhs;
+    return *this; 
+}
+
+FloatType& FloatType::multiply(float rhs) 
+{
+    if (value != nullptr) *value *= rhs;
+    return *this;
+}
+
+FloatType& FloatType::divide(float rhs)
+{
+    if (rhs == 0.0f) 
+    {
+        std::cout << "warning: floating point division by zero!" << std::endl;
+    }
+    if (value != nullptr) *value /= rhs;
```

Comment:
pick a style with regard to your use of `if( ) single line expression` 
choices:
- use curly braces
- don't use curly braces but put the single-line expression on the next line
- don't use curly braces but put the single-line expression on the same line.
- [ ] revise